src/callout.c

```diff
@@ -131,139 +88,68 @@ int timer_nextTimer(void) {
-                delay -= ptr->time; node->time = delay;
-                prev = ptr;
-                ptr = ptr->next;
-            }
-            i++;
+        // chase the queue looking for the right place. 
+        for ( i++; ptr->next && node->time >= ptr->next->time; ptr = ptr->next ) i++;
+        if ( ptr == queue && node->time < ptr->time ) {
+           // Start of queue, insert.
+           queue = node; node->next = ptr;
+        } else if ( ptr->next ) {
+           // Middle of queue, insert
+           node->next = ptr->next; ptr->next = node;
+        } else {
+           // End of queue, append.
+           ptr->next = node;
         }
-        prev->next = node;
     }
-    my_log(LOG_DEBUG, 0, "Created timeout %d (#%d) - delay %d secs",
-            node->id, i, node->time);
     debugQueue();
-
+    my_log(LOG_DEBUG, 0, "Created timeout %d (#%d) - delay %d secs", node->id, i, delay);
     return node->id;
 }
 
 /**
 *   returns the time until the timer is scheduled
 */
 int timer_leftTimer(int timer_id) {
-    struct timeOutQueue *ptr;
-    int left = 0;
-
-    if (!timer_id)
-        return -1;
-
-    for (ptr = queue; ptr; ptr = ptr->next) {
-        left += ptr->time;
-        if (ptr->id == timer_id) {
-            return left;
-        }
+    struct timeOutQueue *ptr = queue;
+    struct timespec curtime;
+
+    if (!timer_id || !queue) return -1;
+    while (ptr && ptr->id != timer_id) ptr = ptr->next;
+    if ( ptr ){
+        clock_gettime (CLOCK_MONOTONIC, &curtime);
+        return (ptr->time - curtime.tv_sec);
     }
     return -1;
 }
 
-/**
-*   clears the associated timer.  Returns 1 if succeeded.
-*/
-int timer_clearTimer(int  timer_id) {
-    struct timeOutQueue  *ptr, *prev;
-    int i = 0;
-
-    if (!timer_id)
-        return 0;
-
-    prev = ptr = queue;
-
-    /*
-     * find the right node, delete it. the subsequent node's time
-     * gets bumped up
-     */
-
-    debugQueue();
-    while (ptr) {
-        if (ptr->id == timer_id) {
-            /* got the right node */
-
-            /* unlink it from the queue */
-            if (ptr == queue)
-                queue = queue->next;
-            else
-                prev->next = ptr->next;
-
-            /* increment next node if any */
-            if (ptr->next != 0)
-                (ptr->next)->time += ptr->time;
-
-            if (ptr->data)
-                free(ptr->data);
-            my_log(LOG_DEBUG, 0, "deleted timer %d (#%d)", ptr->id, i);
-            free(ptr);
-            debugQueue();
-            return 1;
-        }
-        prev = ptr;
-        ptr = ptr->next;
-        i++;
-    }
-    // If we get here, the timer was not deleted.
-    my_log(LOG_DEBUG, 0, "failed to delete timer %d (#%d)", timer_id, i);
-    debugQueue();
-    return 0;
-}
-
 /**
  * debugging utility
  */
 static void debugQueue(void) {
-    struct timeOutQueue  *ptr;
+    struct timeOutQueue  *ptr; int i = 1;
 
     for (ptr = queue; ptr; ptr = ptr->next) {
```

Comment:
as `i` is used only in this `for`-loop it should be initialized in `for` and also incremented here:
`for (ptr = queue, i = 1; ptr; ptr = ptr->next, i++)`